internal/sql/driver/result.go

```diff
@@ -47,16 +49,28 @@ type QueryResult struct {
 }
 
 // NewQueryResult creates a new QueryResult.
-func NewQueryResult(qid sql.QueryID, md sql.RowMetadata, page *sql.Page, ss *SQLService, conn *icluster.Connection, cbs int32) (*QueryResult, error) {
+func NewQueryResult(ctx context.Context, qid sql.QueryID, md sql.RowMetadata, page *sql.Page, ss *SQLService, conn *icluster.Connection, cbs int32) (*QueryResult, error) {
+	doneCh := make(chan struct{})
 	qr := &QueryResult{
 		queryID:          qid,
 		metadata:         md,
 		page:             page,
 		ss:               ss,
 		conn:             conn,
+		doneCh:           doneCh,
 		cursorBufferSize: cbs,
 		index:            0,
 	}
+	// the following goroutine cancels the query when the context is canceled.
+	go func() {
+		select {
+		case <-ctx.Done():
+			// ignoring the error here, since there is nothing to do on error.
+			_ = qr.Close()
+		case <-doneCh:
```

Comment:
when we close doneCh we execute this case right?  and `return` shuts down this goroutine

Reply:
That's right.